Separate resource submission from resource save

cdosborn opened the pull request:
## Description

This pr separates request submission from the saving of resources like quota/allocation. It also improves the experience of approving/denying requests because it was sending emails that didn't make sense.

There are several large commits. Whenever a lot of related changes are made I make an effort to include large commit messages. Viewing it per commit, should be ideal.

Gif quality is beta.
![nov 7 2017 1-49 pm](https://user-images.githubusercontent.com/3847314/32517519-eec8ba9a-c3c3-11e7-93f5-7cee5babc358.gif)

Related atmosphere pr https://github.com/cyverse/atmosphere/pull/541.

## Checklist before merging Pull Requests
- [ ] New test(s) included to reproduce the bug/verify the feature
- [ ] Documentation created/updated at [Example link to documentation](https://example.test/doc#new_section) to give context to the feature
- [ ] Reviewed and approved by at least one other contributor.
- [ ] New variables supported in Clank
- [ ] New variables committed to secrets repos

lenards commented on troposphere/static/js/components/admin/ResourceRequest/ResourceRequest.jsx:
```diff
@@ -36,148 +38,133 @@ export default React.createClass({
             identities = stores.IdentityStore.fetchWhere({
                 username: request.get("created_by").username
             });
+        }
```
Thank you for wrapping this in `{...}`s.

lenards commented on troposphere/static/js/components/admin/ResourceRequest/ResourceRequest.jsx:
```diff
@@ -36,148 +38,133 @@ export default React.createClass({
     },
 
-    onDeny(view) {
+    onDeny(reason) {
```
It does not appear that `reason` is used. Curious if that is intentional (it likely is).

cdosborn replied:
welp that is a bug, good eye

lenards replied:
👍

lenards commented on troposphere/static/js/components/admin/ResourceRequest/ResourceRequest.jsx:
```diff
@@ -36,148 +38,133 @@ export default React.createClass({
+            .then(() => {
+                browserHistory.push("/application/admin/resource-requests");
+            })
+            .always(this.setActionResolved);
```
Would this be where we're include `reason`? (prior to call the action)

lenards commented on troposphere/static/js/components/admin/ResourceRequest/ResourceRequestView.jsx:
```diff
@@ -96,68 +45,116 @@ export default React.createClass({
+
+    renderIdentitiesSection() {
+        let { identities, onIdentitySave: onSave } = this.props;
+        let { section } = this.style();
```
Just want to note for folks _watching_ (and for me):
 ... the statement that appears line 72 here is I think [a variation of destructuring they call "unpacking"](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) where the value of `onIdentitySave` from `this.props` is _bound_ to `onSave`.

You'll see that `onSave` is later used in the `.map(...)`.

lenards commented:
I think this is _good to go_. 👍 

I **am** curious about whether `reason` should be passed as a status or additional data when doing an `updateRequest` (this that's the action name).

steve-gregory commented on troposphere/static/js/actions/errorHandler.js:
```diff
@@ -0,0 +1,30 @@
+
+    // Note: this error handler supports jQuery style promises. When a jQuery
+    // promise is rejected/fails, it calls the error handler with a jqXHR, so
+    // here we are anticipating that response is a jqXHR
```
Handy! I could see this being used elsewhere 👍

steve-gregory commented on troposphere/static/js/components/admin/ResourceRequest/AllocationSourceView.jsx:
```diff
@@ -0,0 +1,186 @@
+
+AllocationEditableSaveableView.propTypes = {
+    onSave: React.PropTypes.func.isRequired
+}
```
First time i've seen this syntax used to define `propTypes` for a class, could you comment on this new usage?

cdosborn replied:
propTypes no longer exist as part of React, but we're using an addon to keep the functionality. When you use es6 classes, propTypes are sought as an attribute on the class.

lenards replied:
@steve-gregory - 

The quick answer is: 

Since `AllocationEditableSaveableView` is defined as an ES6 (aka ES2015) class (instead of `React.createClass({...})`), this is how `propTypes` are defined (given you're not passing in the object literal to `createClass(..)`.

Also, `propTypes` has moved into it's own module (out of the `React` module). So there is a full blown "module" called `PropTypes`. Facebook has provided a codemod that will do some of the translation. 

Why extend `React.Component`? 

The traditional inheritance that Connor leverages here is quite elegant (in my opinion). So, I think this is a great example of where you'd use this over a regular prototypal definition.

cdosborn commented:
Pr is ready for review. I fixed several issues. 

Calling collection.set() didn't behave like I had thought. I also removed calling isMounted in my feature using cancellables instead, because setting state like ismounted, doesn't work.

lenards commented:
I'm not sure why Travis is showing this in "yellow"/warning. The build was complete, clean ✅

lenards commented:
there! ✅
